display/SINK.print/print_callbacks.cpp

```diff
@@ -0,0 +1,178 @@
+            }
+        }
+    } else {
+        
+        nlohmann::json j;
+        j["units"] = {{"time", "ns"}, {"size", "bytes"}};
+        
+        if (data->display_metadata)
+            j["metadata"] = data->metadata;
+        
+        if (data->display_compact) {
+            for (auto& [level,host]: data->host)
+                j["host"][level] = json_compact(host);
+
+            if (!data->device.empty())
+                j["device"] = json_compact(data->device);
+
+            for (auto& [level,traffic]: data->traffic)
+                j["traffic"][level] = json_compact(traffic);
+
+        } else {
+            for (auto& [level,host]: data->host)
+                j["host"][level] = json_extented(host, std::make_tuple("Hostname", "Process", "Thread"));
+                
+            if (!data->device.empty())
+                j["device"] = json_extented(data->device,std::make_tuple(
+                    "Hostname", "Process","Thread", "Device pointer","Subdevice pointer"));
+
+            for (auto& [level,traffic]: data->traffic)
+                    j["traffic"][level] = json_extented(traffic,std::make_tuple(
+                        "Hostname", "Process", "Thread"));
+        }
+        std::cout << j << std::endl;
+    }
+
+    /* Desolate user data */
+    delete data;
+}
+
+void aggreg_host_usr_callbacks(
+    common_data_t *common_data, void *usr_data, const char* hostname, int64_t vpid, 
+    uint64_t vtid, int64_t backend, const char* name, uint64_t dur, uint64_t min, 
+    uint64_t max, uint64_t count, uint64_t err
+)
+{
+    tally_data_t *data = (tally_data_t *) usr_data;
+
+    TallyCoreTime a{dur, err};
+    const int level = backend_level[backend];
+    data->host_backend_name[level].insert(backend_name[backend]);
+    data->host[level][hpt_function_name_t(hostname, vpid, vtid, name)] += a;
+}
+
+void aggreg_kernel_usr_callbacks(
+    common_data_t *common_data, void *usr_data, const char* hostname, int64_t vpid,
+    uint64_t vtid, int64_t backend, const char* name, const char* metadata, 
+    uint64_t did, uint64_t sdid, uint64_t dur, uint64_t min, uint64_t max, 
+    uint64_t count, uint64_t err
+)
+{
+    tally_data_t *data = (tally_data_t *) usr_data;
+
+    const auto name_demangled = (data->demangle_name) ? f_demangle_name(name) : name;
+    const auto name_with_metadata = (data->display_kernel_verbose && !strcmp(metadata, "")) ? name_demangled + "[" + metadata + "]" : name_demangled;
+
+    TallyCoreTime a{dur, err};
+    data->device[hpt_device_function_name_t(hostname, vpid, vtid, did, sdid, name_with_metadata)] += a;
+}
+
+void aggreg_traffic_usr_callbacks(
+    common_data_t *common_data, void *usr_data, const char* hostname, int64_t vpid,
+    uint64_t vtid, int64_t backend, const char* name, uint64_t size, uint64_t min,
+    uint64_t max, uint64_t count, uint64_t err
+)
+{
+    tally_data_t *data = (tally_data_t *) usr_data;
+
+    TallyCoreByte a{(uint64_t)size, false};
```

Comment:
why you need to case?

Reply:
You are right, I copy and pasted the code, so I need to be careful since here there are things that are not needed any more.